src/core/observables/ParticleTraits.hpp

```diff
@@ -0,0 +1,62 @@
+/*
+ * Copyright (C) 2010-2020 The ESPResSo project
+ *
+ * This file is part of ESPResSo.
+ *
+ * ESPResSo is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * ESPResSo is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+#ifndef OBSERVABLES_PARTICLE_TRAITS
+#define OBSERVABLES_PARTICLE_TRAITS
+
+#include "Particle.hpp"
+#include "config.hpp"
+
+namespace ParticleObservables {
+/**
+ * Template specialization for `Particle`. The traits mechanism is used to get
+ * indirect access to particle properties. This helps making the implementation
+ * of observables independent of the particle type.
+ */
+template <> struct traits<Particle> {
+  auto position(Particle const &p) const { return p.r.p; }
+  auto velocity(Particle const &p) const { return p.m.v; }
+  auto mass(Particle const &p) const {
+#ifdef VIRTUAL_SITES
+    // we exclude virtual particles since their mass does not have a meaning
+    if (p.p.is_virtual)
+      return decltype(p.p.mass){};
+#endif
+    return p.p.mass;
+  }
+  auto charge(Particle const &p) const { return p.p.q; }
+  auto force(Particle const &p) const {
```

Comment:
I don't think that the force should be here, because it is not a property of a single but of a set of particles.

Reply:
how is the force not a property of a single particle?

Reply:
Because forces stem from interactions with other particles. There is another, more technical, difference, the forces depend on the system parameters while all the other properties do not.
It'd be nice if this was encoded in the type...

Reply:
> Because forces stem from interactions with other particles

Like velocities may.

Reply:
I don't agree with this. From a given state point you can tell the velocity by only using the state of that particle, which is not true for forces. More fundamentally a particles does not 'have' a force,
that is a property of the system (e.g. of the Hamiltonian), otherwise physics would be pretty boring.

Reply:
To put it differently, you can chose the velocity (and position for that matter) for any particle independently, but given a state of the system and a Hamiltonian, there is only one 'correct' set
of forces for particles, and in general this set depends on the set of the positions and velocities of all particles.

Reply:
That is even more complicate because we usually have a Hamiltonian where forces can depend on velocities... So what do you want to propose now?

Reply:
That does not change anything. A force particle trait is not composable, because it can not be calculated per particle. That there is something related to forces stored on the particles is an
implementation detail, and leaking it leads to all kind of problems.

Reply:
(Btw., I don't think that we can have a "Hamiltonian where forces can depend on velocities")